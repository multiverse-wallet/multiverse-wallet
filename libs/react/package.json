{
  "name": "@multiverse-wallet/react",
  "version": "0.0.3"
}
