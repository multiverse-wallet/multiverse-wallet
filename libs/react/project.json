{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/react/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "build": {
      "executor": "@nrwl/web:rollup",
      "outputs": [
        "{options.outputPath}"
      ],
      "options": {
        "outputPath": "dist/libs/react",
        "tsConfig": "libs/react/tsconfig.lib.json",
        "project": "libs/react/package.json",
        "entryFile": "libs/react/src/index.ts",
        "external": [
          "react/jsx-runtime"
        ],
        "rollupConfig": "@nrwl/react/plugins/bundle-rollup",
        "compiler": "babel",
        "assets": [
          {
            "glob": "libs/react/README.md",
            "input": ".",
            "output": "."
          }
        ],
        "updateBuildableProjectDepsInPackageJson": true,
        "buildableProjectDepsInPackageJsonType": "dependencies"
      },
      "dependsOn": ["multiverse"]
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": [
        "{options.outputFile}"
      ],
      "options": {
        "lintFilePatterns": [
          "libs/react/**/*.{ts,tsx,js,jsx}"
        ]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": [
        "coverage/libs/react"
      ],
      "options": {
        "jestConfig": "libs/react/jest.config.ts",
        "passWithNoTests": true
      }
    }
  }
}
