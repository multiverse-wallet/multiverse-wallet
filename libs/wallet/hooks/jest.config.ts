/* eslint-disable */
export default {
  displayName: 'wallet-hooks',
  preset: '../../../jest.preset.js',
  transform: {
    '^.+\\.[tj]sx?$': 'babel-jest'
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../../coverage/libs/wallet/hooks'
};
