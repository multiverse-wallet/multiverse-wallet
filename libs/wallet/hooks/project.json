{
  "$schema": "../../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/wallet/hooks/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": [
        "{options.outputFile}"
      ],
      "options": {
        "lintFilePatterns": [
          "libs/wallet/hooks/**/*.{ts,tsx,js,jsx}"
        ]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": [
        "coverage/libs/wallet/hooks"
      ],
      "options": {
        "jestConfig": "libs/wallet/hooks/jest.config.ts",
        "passWithNoTests": true
      }
    }
  }
}
