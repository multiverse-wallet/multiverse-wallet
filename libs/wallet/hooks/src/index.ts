export * from './lib/use-wallet-state';
export * from './lib/use-well-known-names';
