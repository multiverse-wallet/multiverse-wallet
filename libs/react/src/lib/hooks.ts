import { useCallback, useEffect, useMemo, useState } from 'react';
import {
  Account,
  Network,
  PublicAPI,
  Transaction,
} from '@multiverse-wallet/multiverse';
import * as xrpl from 'xrpl';

const defaultApi = new PublicAPI();

export interface Opts {
  // Overrides the API instance to be used by the hook.
  overrideApi?: PublicAPI;
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface UseConnectOpts extends Opts {}

export interface IUseConnect {
  isActive: boolean;
  isConnected: boolean;
  connect: () => void;
  disconnect: () => void;
  error?: Error;
  revalidate: () => void;
}

export function useConnect({ overrideApi }: UseConnectOpts = {}): IUseConnect {
  const api = useMemo(() => overrideApi || defaultApi, [overrideApi]);
  const [isActive, setIsActive] = useState(false);
  const [isConnected, setIsConnected] = useState(false);
  const [error, setError] = useState<Error>();
  const connect = () => {
    api
      ?.connect()
      .then(() => setError(undefined))
      .catch((e) => setError(e));
  };
  const disconnect = () => {
    api
      ?.disconnect()
      .then(() => setError(undefined))
      .catch((e) => setError(e));
  };
  const revalidate = useCallback(() => {
    api
      ?.isConnected()
      .then((c) => {
        setError(undefined);
        setIsConnected(c);
      })
      .catch((e) => setError(e));
    api
      ?.isActive()
      .then((a) => {
        setError(undefined);
        setIsActive(a);
      })
      .catch((e) => setError(e));
  }, [api]);
  useEffect(() => {
    revalidate();
    return api?.on('accountChanged', () => {
      revalidate();
    });
  }, [api, revalidate]);
  useEffect(() => {
    revalidate();
    return api?.on('connectionChanged', () => {
      revalidate();
    });
  }, [api, revalidate]);
  useEffect(() => {
    revalidate();
    return api?.on('update', () => {
      revalidate();
    });
  }, [api, revalidate]);
  return {
    isConnected,
    isActive,
    connect,
    disconnect,
    error,
    revalidate,
  };
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface UseAccountOpts extends Opts {}

export interface IUseAccount {
  account?: Account;
  error?: Error;
  revalidate: () => void;
}

export function useAccount({ overrideApi }: UseAccountOpts = {}): IUseAccount {
  const api = useMemo(() => overrideApi || defaultApi, [overrideApi]);
  const [account, setAccount] = useState<Account>();
  const [error, setError] = useState<Error>();
  const revalidate = useCallback(() => {
    api
      ?.getAccount()
      .then((a) => {
        setError(undefined);
        setAccount(a);
      })
      .catch((e) => setError(e));
  }, [api]);
  useEffect(() => {
    revalidate();
    return api?.on('accountChanged', () => {
      revalidate();
    });
  }, [api, revalidate]);
  return {
    account,
    error,
    revalidate,
  };
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface UseNetworkOpts extends Opts {}

export interface IUseNetwork {
  network?: Network;
  error?: Error;
  revalidate: () => void;
}

export function useNetwork({ overrideApi }: UseNetworkOpts = {}): IUseNetwork {
  const api = useMemo(() => overrideApi || defaultApi, [overrideApi]);
  const [network, setNetwork] = useState<Network>();
  const [error, setError] = useState<Error>();
  const revalidate = useCallback(() => {
    api
      ?.getNetwork()
      .then((n) => {
        setError(undefined);
        setNetwork(n);
      })
      .catch((e) => setError(e));
  }, [api]);
  useEffect(() => {
    revalidate();
    return api?.on('networkChanged', () => {
      revalidate();
    });
  }, [api, revalidate]);
  return {
    network,
    error,
    revalidate,
  };
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface UseTransactionOpts extends Opts {
  transactionId?: string;
}

export interface IUseTransaction {
  requestTransaction: (txJson: unknown) => void;
  transaction?: Transaction;
  error?: Error;
  revalidate: () => void;
}

export function useTransaction({
  transactionId: defaultTransactionId,
  overrideApi,
}: UseTransactionOpts = {}): IUseTransaction {
  const api = useMemo(() => overrideApi || defaultApi, [overrideApi]);
  const [transactionId, setTransactionId] = useState<string | undefined>(
    defaultTransactionId
  );
  const [transaction, setTransaction] = useState<Transaction>();
  const [error, setError] = useState<Error>();
  const requestTransaction = useCallback(
    (transaction: unknown) => {
      return api
        ?.requestTransaction(transaction)
        .then((id) => {
          setError(undefined);
          setTransactionId(id);
        })
        .catch((e) => setError(e));
    },
    [api]
  );
  const revalidate = useCallback(() => {
    !!transactionId &&
      api
        ?.getTransaction(transactionId)
        .then((t) => {
          setError(undefined);
          setTransaction(t);
        })
        .catch((e) => setError(e));
  }, [transactionId, api]);
  useEffect(() => {
    revalidate();
    return api?.on('transactionStatusChanged', (tx) => {
      revalidate();
    });
  }, [api, revalidate]);
  return {
    transaction,
    error,
    revalidate,
    requestTransaction,
  };
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface UseClientOpts extends Opts {
  clientOptions?: xrpl.ClientOptions;
}

export function useClient({ overrideApi, clientOptions }: UseClientOpts) {
  const { network } = useNetwork({ overrideApi });
  const [client, setClient] = useState<xrpl.Client>();
  const [error, setError] = useState<Error>();
  const reconnect = useCallback(async () => {
    setClient(undefined);
    if (!network) return;
    const client = new xrpl.Client(network.server, clientOptions);
    await client.connect().catch((e) => setError(e));
    client.on('error', (e) => setError(e));
    client.on('disconnected', () => reconnect());
    setClient(client);
    setError(undefined);
  }, [network, clientOptions]);
  const disconnect = useCallback(() => {
    client?.disconnect();
  }, [client]);
  useEffect(() => {
    reconnect();
    return () => disconnect();
  }, [reconnect, disconnect]);
  return {
    client,
    error,
  };
}
