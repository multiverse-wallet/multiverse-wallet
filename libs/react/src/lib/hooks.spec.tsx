import {
  GetTransactionRequest,
  MockTransport,
  PublicAPI,
  RPCRequest,
} from "@multiverse-wallet/multiverse";
import { render, act } from "@testing-library/react";
import { useConnect, useAccount, useNetwork, useTransaction } from "./hooks";

describe("useConnect", () => {
  let api: PublicAPI;
  let mockTransport: MockTransport;
  beforeEach(() => {
    api = new PublicAPI();
    mockTransport = new MockTransport();
    api.setTransport(mockTransport);
  });
  const TestComponent = () => {
    const { isConnected } = useConnect({ overrideApi: api });
    return <div>{isConnected ? "Connected" : "Not Connected"}</div>;
  };
  it("should render successfully", async () => {
    let baseElement;
    await act(() => {
      const component = render(<TestComponent />);
      baseElement = component.baseElement;
    });
    expect(baseElement).toBeTruthy();
  });
  it("should render connected", async () => {
    let baseElement;
    await act(() => {
      mockTransport.mockResolvedValue(true);
      const component = render(<TestComponent />);
      baseElement = component.baseElement;
    });
    expect(baseElement).toMatchSnapshot();
  });
  it("should render not connected", async () => {
    let baseElement;
    await act(() => {
      mockTransport.mockResolvedValue(false);
      const component = render(<TestComponent />);
      baseElement = component.baseElement;
    });
    expect(baseElement).toMatchSnapshot();
  });
});

describe("useAccount", () => {
  let api: PublicAPI;
  let mockTransport: MockTransport;
  beforeEach(() => {
    api = new PublicAPI();
    mockTransport = new MockTransport();
    api.setTransport(mockTransport);
  });
  const TestComponent = () => {
    const { account, error } = useAccount({ overrideApi: api });
    if (error) {
      return <div>{error?.message}</div>;
    }
    return <div>{account?.name}</div>;
  };
  it("should render successfully", async () => {
    let baseElement;
    await act(() => {
      const component = render(<TestComponent />);
      baseElement = component.baseElement;
    });
    expect(baseElement).toBeTruthy();
  });
  it("should render account", async () => {
    let baseElement;
    await act(() => {
      mockTransport.mockResolvedValue({ name: "Test Account" });
      const component = render(<TestComponent />);
      baseElement = component.baseElement;
    });
    expect(baseElement).toMatchSnapshot();
  });
  it("should render error", async () => {
    let baseElement;
    await act(() => {
      mockTransport.mockRejectedValue(new Error("an error occurred"));
      const component = render(<TestComponent />);
      baseElement = component.baseElement;
    });
    expect(baseElement).toMatchSnapshot();
  });
});

describe("useNetwork", () => {
  let api: PublicAPI;
  let mockTransport: MockTransport;
  beforeEach(() => {
    api = new PublicAPI();
    mockTransport = new MockTransport();
    api.setTransport(mockTransport);
  });
  const TestComponent = () => {
    const { network, error } = useNetwork({ overrideApi: api });
    if (error) {
      return <div>{error?.message}</div>;
    }
    return <div>{network?.name}</div>;
  };
  it("should render successfully", async () => {
    let baseElement;
    await act(() => {
      const component = render(<TestComponent />);
      baseElement = component.baseElement;
    });
    expect(baseElement).toBeTruthy();
  });
  it("should render network", async () => {
    let baseElement;
    await act(() => {
      mockTransport.mockResolvedValue({ name: "Test Network" });
      const component = render(<TestComponent />);
      baseElement = component.baseElement;
    });
    expect(baseElement).toMatchSnapshot();
  });
  it("should render error", async () => {
    let baseElement;
    await act(() => {
      mockTransport.mockRejectedValue(new Error("an error occurred"));
      const component = render(<TestComponent />);
      baseElement = component.baseElement;
    });
    expect(baseElement).toMatchSnapshot();
  });
});

describe("useTransaction", () => {
  let api: PublicAPI;
  let mockTransport: MockTransport;
  beforeEach(() => {
    api = new PublicAPI();
    mockTransport = new MockTransport();
    api.setTransport(mockTransport);
  });
  const transactionId = "test"
  const TestComponent = () => {
    const { transaction, error } = useTransaction({
      transactionId,
      overrideApi: api,
    });
    if (error) {
      return <div>{error?.message}</div>;
    }
    return <div>{transaction?.id}</div>;
  };
  it("should render successfully", async () => {
    let baseElement;
    await act(() => {
      const component = render(<TestComponent />);
      baseElement = component.baseElement;
    });
    expect(baseElement).toBeTruthy();
  });
  it("should render transaction", async () => {
    let baseElement;
    await act(() => {
      mockTransport.mockResolvedValue(
        (req: RPCRequest<GetTransactionRequest>) => {
          return {
            [transactionId]: {
              id: transactionId,
            },
          }[req.data.id];
        }
      );
      const component = render(<TestComponent />);
      baseElement = component.baseElement;
    });
    expect(baseElement).toMatchSnapshot();
  });
  it("should render error", async () => {
    let baseElement;
    await act(() => {
      mockTransport.mockRejectedValue(new Error("an error occurred"));
      const component = render(<TestComponent />);
      baseElement = component.baseElement;
    });
    expect(baseElement).toMatchSnapshot();
  });
});
