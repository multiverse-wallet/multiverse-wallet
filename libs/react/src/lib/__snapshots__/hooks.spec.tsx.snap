// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useAccount should render account 1`] = `
<body>
  <div>
    <div>
      Test Account
    </div>
  </div>
</body>
`;

exports[`useAccount should render error 1`] = `
<body>
  <div>
    <div>
      an error occurred
    </div>
  </div>
</body>
`;

exports[`useConnect should render connected 1`] = `
<body>
  <div>
    <div>
      Connected
    </div>
  </div>
</body>
`;

exports[`useConnect should render not connected 1`] = `
<body>
  <div>
    <div>
      Not Connected
    </div>
  </div>
</body>
`;

exports[`useNetwork should render error 1`] = `
<body>
  <div>
    <div>
      an error occurred
    </div>
  </div>
</body>
`;

exports[`useNetwork should render network 1`] = `
<body>
  <div>
    <div>
      Test Network
    </div>
  </div>
</body>
`;

exports[`useTransaction should render error 1`] = `
<body>
  <div>
    <div>
      an error occurred
    </div>
  </div>
</body>
`;

exports[`useTransaction should render transaction 1`] = `
<body>
  <div>
    <div>
      test
    </div>
  </div>
</body>
`;
